planner/core/exhaust_physical_plans.go

```diff
@@ -2227,20 +2227,92 @@ func (la *LogicalAggregation) getStreamAggs(prop *property.PhysicalProperty) []P
 	return streamAggs
 }
 
+// TODO: support more operators and distinct later
+func (la *LogicalAggregation) checkCanPushDownToMPP() bool {
+	for _, agg := range la.AggFuncs {
+		if agg.HasDistinct {
+			return false
+		}
+		switch agg.Name {
+		case ast.AggFuncSum, ast.AggFuncMin, ast.AggFuncCount, ast.AggFuncMax, ast.AggFuncFirstRow, ast.AggFuncAvg:
+		default:
+			return false
+		}
+		if !expression.CanExprsPushDown(la.SCtx().GetSessionVars().StmtCtx, agg.Args, la.SCtx().GetClient(), kv.TiFlash) {
+			return false
+		}
+	}
+	return true
+}
+
+func (la *LogicalAggregation) tryToGetMppHashAggs(prop *property.PhysicalProperty) (hashAggs []PhysicalPlan) {
+	if !prop.IsEmpty() {
+		return nil
+	}
+	if prop.TaskTp != property.RootTaskType && prop.TaskTp != property.MppTaskType {
+		return
+	}
+	if prop.PartitionTp == property.BroadcastType {
+		return
+	}
+	partitionCols := la.GetGroupByCols()
+	if len(partitionCols) != 0 {
+		if prop.PartitionTp == property.HashType {
+			if matches := prop.IsSubsetOf(partitionCols); len(matches) != 0 {
+				partitionCols = chooseSubsetOfJoinKeys(partitionCols, matches)
+			} else {
+				// do not satisfy the property of its parent, so return empty
+				return
+			}
+		}
+		// TODO: permute various partition columns from group-by columns
+		// 1-phase
+		childProp := &property.PhysicalProperty{TaskTp: property.MppTaskType, ExpectedCnt: math.MaxFloat64, PartitionTp: property.HashType, PartitionCols: partitionCols, CanAddEnforcer: true}
+		agg := NewPhysicalHashAgg(la, la.stats.ScaleByExpectCnt(prop.ExpectedCnt), childProp)
+		agg.SetSchema(la.schema.Clone())
+		agg.MppRunMode = Mpp1Phase
+		hashAggs = append(hashAggs, agg)
+		// 2-phase
+		childProp = &property.PhysicalProperty{TaskTp: property.MppTaskType, ExpectedCnt: math.MaxFloat64, PartitionTp: property.AnyType}
+		agg = NewPhysicalHashAgg(la, la.stats.ScaleByExpectCnt(prop.ExpectedCnt), childProp)
+		agg.SetSchema(la.schema.Clone())
+		agg.MppRunMode = Mpp2Phase
+		agg.PartitionCols = partitionCols
+		hashAggs = append(hashAggs, agg)
+
+		if prop.TaskTp == property.RootTaskType {
```

Comment:
If partition cols is not nil, the task type can only be hash type.

Reply:
here the partition cols are not nil.